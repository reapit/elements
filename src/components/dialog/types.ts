import { FC, HTMLAttributes } from 'react'

import { ElDialogContainerProps } from './style'

export interface DialogContainerProps extends ElDialogContainerProps {}

export interface DialogTitleProps extends HTMLAttributes<HTMLHeadingElement> {}

export interface DialogBodyProps extends HTMLAttributes<HTMLDivElement> {}

export interface DialogFooterProps extends HTMLAttributes<HTMLDivElement> {}

export type DialogFooterItems = FC<{ closeDialog: VoidFunction }>
