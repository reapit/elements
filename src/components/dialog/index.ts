export * from './dialog'
export * from './types'
export * from './use-dialog-appearance'
