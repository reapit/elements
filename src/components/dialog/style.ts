import { DialogHTMLAttributes } from 'react'

import { styled } from '@linaria/react'

export type SmallDialogSize = 'small'
export type MediumDialogSize = 'medium'

// Extended version of the `DialogHTMLAttributes` interface, which includes
// the 'data-size' attribute that used as the controller of the dialog size.
export type ElDialogContainerProps = DialogHTMLAttributes<HTMLDialogElement> & {
  /**
   * The size of the dialog. The small size is used by default.
   * - 'small' will render the dialog in a small size
   * - 'medium' will render the dialog in a medium size
   *
   * @default 'small'
   */
  'data-size'?: SmallDialogSize | MediumDialogSize
}

export const ElDialogContainer = styled.dialog<ElDialogContainerProps>`
  // NOTE: private CSS variables that deliberated used by the dialog component
  --__dialog-box-shadow: 0px 4px 16px 0px rgba(96, 120, 144, 0.16);

  z-index: 99;
  overflow: auto;
  border: none;
  box-shadow: var(--__dialog-box-shadow);
  border-radius: var(--corner-radius-corner-lg, 8px);
  padding: var(--spacing-spacing-5, 20px) var(--spacing-spacing-6, 24px);
  min-width: 300px;
  max-height: 100%;

  // calculate the width of the dialog based on the size of the container
  width: calc(100% - var(--unit-16, 64px));
  max-width: 20rem;

  &[data-size='medium'] {
    width: calc(100% - var(--unit-8, 32px));
    max-width: 30rem;
  }

  &::backdrop {
    opacity: 0.5;
    animation: dialogBackdropFadeIn 225ms cubic-bezier(0.4, 0, 0.2, 1);
    background: var(--black);
  }

  &[open] {
    animation: dialogContainerFadeIn 225ms ease-in forwards;
  }

  @keyframes dialogContainerFadeIn {
    from {
      opacity: 0;
    }
    to {
      opacity: 1;
    }
  }

  @keyframes dialogBackdropFadeIn {
    from {
      opacity: 0;
    }
    to {
      opacity: 0.5;
    }
  }
`

/**
 * HTML element that renders the title of the dialog
 * It is intended to be used as a child of the `Dialog` component.
 * it doesn't have the maximum lines or truncation.
 */
export const ElDialogTitle = styled.h2`
  color: var(--text-primary, #222b33);
  font-family: var(--font-family, Inter);
  font-size: var(--font-size-lg, 18px);
  font-style: normal;
  font-weight: 600;
  line-height: var(--line-height-lg, 24px);
  letter-spacing: var(--letter-spacing-lg, -0.18px);
  padding-bottom: var(--spacing-4, 12px);
  word-wrap: break-word;
`

/**
 * HTML element that renders the body of the dialog
 */
export const ElDialogBody = styled.div`
  font-size: var(--font-size-default);
`

export const ElDialogFooter = styled.div`
  display: flex;
  justify-content: flex-end;
  padding-top: var(--spacing-5, 20px);
`
