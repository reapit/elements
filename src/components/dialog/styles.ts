import { DialogHTMLAttributes } from 'react'

import { styled } from '@linaria/react'

export type SmallDialogSize = 'small'
export type MediumDialogSize = 'medium'

// Extended version of the `DialogHTMLAttributes` interface, which includes
// the 'data-size' attribute that used as the controller of the dialog size.
type ElDialogContainerProps = DialogHTMLAttributes<HTMLDialogElement> & {
  /**
   * The size of the dialog. The small size is used by default.
   * - 'small' will render the dialog in a small size
   * - 'medium' will render the dialog in a medium size
   *
   * @default 'small'
   */
  'data-size'?: SmallDialogSize | MediumDialogSize
}

export const ElDialog = styled.dialog<ElDialogContainerProps>`
  // NOTE: private CSS variables that deliberated used by the dialog component
  --__dialog-box-shadow: 0px 4px 16px 0px rgba(96, 120, 144, 0.16);

  z-index: 99;
  overflow: auto;
  border: none;
  box-shadow: var(--__dialog-box-shadow);
  border-radius: var(--corner-radius-corner-lg, 8px);
  padding: var(--spacing-spacing-5, 20px) var(--spacing-spacing-6, 24px);
  min-width: 300px;
  max-height: 100%;

  // calculate the width of the dialog based on the size of the container
  width: calc(100% - var(--unit-16, 64px));
  max-width: 20rem;

  animation: dialogFadeOut 0.2s ease-out;

  &[data-size='medium'] {
    width: calc(100% - var(--unit-8, 32px));
    max-width: 30rem;
  }

  &[open] {
    animation: dialogFadeIn 0.2s ease-out forwards;

    &::backdrop {
      background-color: var(--black);
      opacity: 0.5;
      animation: dialogBackdropFadeIn 225ms cubic-bezier(0.4, 0, 0.2, 1);
    }
  }

  @keyframes dialogFadeIn {
    from {
      opacity: 0;
      display: none;
    }

    to {
      opacity: 1;
      display: block;
    }
  }

  @keyframes dialogFadeOut {
    from {
      opacity: 1;
      display: block;
    }

    to {
      opacity: 0;
      display: none;
    }
  }

  @keyframes dialogBackdropFadeIn {
    0% {
      background-color: var(--black);
      opacity: 0;
    }

    100% {
      background-color: var(--black);
      opacity: 0.5;
    }
  }
`

/**
 * HTML element that renders the title of the dialog
 * It is intended to be used as a child of the `ElDialogHeader` component.
 * it doesn't have the maximum lines or truncation.
 */
export const ElDialogTitle = styled.h2`
  color: var(--text-primary, #222b33);
  font-family: var(--font-family, Inter);
  font-size: var(--font-size-lg, 18px);
  font-style: normal;
  font-weight: var(--font-weight-semibold);
  line-height: var(--line-height-lg, 24px);
  letter-spacing: var(--letter-spacing-lg, -0.18px);
`
/**
 * HTML element that renders the header of the dialog
 */
export const ElDialogHeader = styled.div`
  padding-bottom: var(--spacing-4, 12px);
  word-wrap: break-word;
`

/**
 * HTML element that renders the body of the dialog
 */
export const ElDialogBody = styled.div`
  font-size: var(--font-size-default);
`

export const ElDialogFooter = styled.div`
  display: flex;
  justify-content: flex-end;
  padding-top: var(--spacing-5, 20px);
`
