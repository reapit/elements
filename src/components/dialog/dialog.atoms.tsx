import { forwardRef } from 'react'

import { DialogBodyProps, DialogFooterProps, DialogContainerProps, DialogTitleProps } from './types'
import { ElDialogBody, ElDialogContainer, ElDialogFooter, ElDialogTitle } from './style'

const Container = forwardRef<HTMLDialogElement, DialogContainerProps>(({ children, ...rest }, ref) => {
  return (
    <ElDialogContainer {...rest} ref={ref}>
      {children}
    </ElDialogContainer>
  )
})

const Title = ({ title, ...rest }: DialogTitleProps) => (
  <ElDialogTitle {...rest} title={title}>
    {title}
  </ElDialogTitle>
)

const Body = ({ children, ...rest }: DialogBodyProps) => <ElDialogBody {...rest}>{children}</ElDialogBody>

const Footer = ({ children, ...rest }: DialogFooterProps) => <ElDialogFooter {...rest}>{children}</ElDialogFooter>

// NOTE: We extend the `Container` component with the `Title`, `Body`, and `Footer` components
// to provide a more convenient way to render the dialog content.
const Dialog = Container as typeof Container & {
  Title: typeof Title
  Body: typeof Body
  Footer: typeof Footer
}

Dialog.Title = Title
Dialog.Body = Body
Dialog.Footer = Footer

export { Dialog }
