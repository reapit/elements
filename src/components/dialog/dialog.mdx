import { Meta, Story, Canvas, Controls } from '@storybook/blocks'
import { RenderHtmlMarkup } from '../../storybook/render-html-markup'
import * as DialogStories from './dialog.stories'

<Meta of={DialogStories} />

# Dialog

The Dialog component that renders a dialog box that overlays the main screen. It is typically used for displaying confirmation or error notifications to users that require an addtional call to action or step.

## Basic Usage

The Dialog is built on top of the `dialog` HTML element and is therefore accessible to screen readers.

To open the dialog, you must use the `showModal` method on the dialog element and avoid using the open prop.

To close the dialog, you must use the `close` method on the dialog element.

By default, the dialog fills it's current container so you will likely wish to place it as high as possible up the DOM tree to achive an overlay effect.

Dialog has two sizes, small and medium, which are applied to the `dialog` element. The small is used by default due to its suitability with mobile-first sizing.

To use the medium size, you can add the `data-size="medium"` prop to the dialog element.

<Canvas of={DialogStories.BasicUsageSmall} story={{ height: '200px' }} />

<RenderHtmlMarkup of={DialogStories.BasicUsageSmall} />

## Medium size

The medium variant is the default size and is suitable for displaying more complex content. It offers a larger viewing area than the small variant.

<Canvas of={DialogStories.BasicUsageMedium} story={{ height: '200px' }} />

<RenderHtmlMarkup of={DialogStories.BasicUsageMedium} />

## React usage

The React component accepts props to toggle the classes that control whether the dialog is opened or not.

Props available for the React version of the component;

<Controls of={DialogStories.ReactUsage} />

<Canvas of={DialogStories.ReactUsage} story={{ height: '300px' }} />
