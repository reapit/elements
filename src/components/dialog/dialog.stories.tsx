import { useState } from 'react'

import type { Meta, StoryObj } from '@storybook/react'
import { action } from '@storybook/addon-actions'

import { Button, ButtonGroup } from '../button'

import { Dialog as DialogReactUsage } from './dialog'
import { Dialog as DialogBasicUsage } from './dialog.atoms'

const meta = {
  title: 'Components/Dialog',
  component: DialogReactUsage,
  parameters: {
    layout: 'padded',
  },
  argTypes: {
    title: {
      control: {
        type: 'text',
      },
    },
    'data-size': {
      control: {
        type: 'select',
        options: ['small', 'medium'],
      },
    },
    isOpen: {
      control: false,
      description: 'Whether the dialog is open or not',
      type: 'boolean',
    },
    children: {
      control: false,
      description: 'JSX content that will placed within the dialog body',
      name: 'content',
      type: 'function',
    },
    footerItems: {
      control: false,
      description: 'The JSX footer content for dialog footer',
      type: 'function',
    },
    handleCloseDialog: action('handleCloseDialog'),
  },
  args: {
    title: 'Welcome to the demo dialog',
    children: "Here's some nice content for the inside of the basic dialog",
    'data-size': undefined,
    isOpen: true,
    handleCloseDialog: action('handleCloseDialog'),
    footerItems: ({ closeDialog }) => (
      <ButtonGroup>
        <Button intent="secondary" onClick={closeDialog}>
          Cancel
        </Button>
        <Button intent="primary" onClick={closeDialog}>
          Save
        </Button>
      </ButtonGroup>
    ),
  },
} satisfies Meta<typeof DialogReactUsage>

export default meta
type Story = StoryObj<typeof meta>

export const BasicUsageSmall: Story = {
  args: {
    isOpen: true,
    title: 'Hello from Small Dialog',
    'data-size': 'small',
  },
  render: ({ title, children, footerItems, isOpen, ...rest }) => {
    return (
      <DialogBasicUsage {...rest} data-size={rest['data-size']} open={isOpen}>
        <DialogBasicUsage.Title title={title} />
        <DialogBasicUsage.Body>{children}</DialogBasicUsage.Body>
        <DialogBasicUsage.Footer>{footerItems({ closeDialog: action('closeDialog') })}</DialogBasicUsage.Footer>
      </DialogBasicUsage>
    )
  },
}

export const BasicUsageMedium: Story = {
  args: {
    isOpen: true,
    title: 'Hello from Medium Dialog',
    'data-size': 'medium',
  },
  render: ({ title, children, footerItems, isOpen, ...rest }) => {
    return (
      <DialogBasicUsage {...rest} data-size={rest['data-size']} open={isOpen}>
        <DialogBasicUsage.Title title={title} />
        <DialogBasicUsage.Body>{children}</DialogBasicUsage.Body>
        <DialogBasicUsage.Footer>{footerItems({ closeDialog: action('closeDialog') })}</DialogBasicUsage.Footer>
      </DialogBasicUsage>
    )
  },
}

export const ReactUsage: Story = {
  args: {
    title: 'Hello from React Usage Dialog',
    children: "Here's some nice content for the inside of the React dialog",
    handleCloseDialog: action('handleCloseDialog'),
  },
  render: ({ title, children, ...rest }) => {
    const [isOpen, setIsOpen] = useState(false)

    return (
      <>
        <ButtonGroup>
          <Button intent="secondary" onClick={() => setIsOpen(true)}>
            Open Dialog
          </Button>
        </ButtonGroup>
        <DialogReactUsage
          {...rest}
          isOpen={isOpen}
          title={title}
          handleCloseDialog={() => setIsOpen(false)}
          footerItems={({ closeDialog }) => (
            <ButtonGroup alignment="right">
              <Button intent="secondary" onClick={closeDialog}>
                Cancel
              </Button>
              <Button intent="primary" onClick={closeDialog}>
                Close
              </Button>
            </ButtonGroup>
          )}
        >
          <p>{children}. Try to change the size of the dialog</p>
        </DialogReactUsage>
      </>
    )
  },
}
