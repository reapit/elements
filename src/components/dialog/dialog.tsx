import { FC, ReactNode } from 'react'

import { DialogContainerProps, DialogFooterItems } from './types'
import { Dialog } from './dialog.atoms'
import { useDialogAppearance } from './use-dialog-appearance'

export interface DialogProps extends DialogContainerProps {
  /**
   * Whether the dialog is open or not.
   *
   * @default false
   */
  isOpen: boolean

  /**
   * Function that has the responsibility to close the dialog.
   */
  handleCloseDialog: () => void

  /**
   * Renders the title of the dialog. It must be a string, and is optional
   */
  title?: string

  /**
   * The content of the dialog, which will be placed within the `Dialog.Body`.
   */
  children?: ReactNode

  /**
   * The footer JSX content of the dialog, which will be placed within the `Dialog.Footer`.
   */
  footerItems: DialogFooterItems
}

/**
 * The `Dialog` component is a wrapper around the `Dialog` atoms that provides
 * a more convenient way to render the dialog content.
 *
 * This component is intended to be used as a controlled component, meaning that it should be
 * used in conjunction with the `useDialog` hook to manage the state of the dialog
 * and ensure that it is properly rendered and styled.
 */
const DialogComposed: FC<DialogProps> = ({
  title,
  footerItems,
  children,
  isOpen = false,
  handleCloseDialog,
  ...props
}) => {
  const ref = useDialogAppearance({ isOpen, onClose: handleCloseDialog })

  return (
    <Dialog {...props} ref={ref}>
      <Dialog.Title title={title} />
      <Dialog.Body>{children}</Dialog.Body>
      <Dialog.Footer>{footerItems({ closeDialog: handleCloseDialog })}</Dialog.Footer>
    </Dialog>
  )
}

export { DialogComposed as Dialog }
