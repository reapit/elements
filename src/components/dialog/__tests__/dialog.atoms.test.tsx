import { render } from '@testing-library/react'
import { Dialog } from '../dialog.atoms'

describe('Dialog.Container', () => {
  it('should match snapshot', () => {
    expect(render(<Dialog />).asFragment()).toMatchSnapshot()
  })
})

describe('Dialog.Title', () => {
  it('should match snapshot', () => {
    expect(render(<Dialog.Title />).asFragment()).toMatchSnapshot()
  })
})

describe('Dialog.Body', () => {
  it('should match snapshot', () => {
    expect(render(<Dialog.Body />).asFragment()).toMatchSnapshot()
  })
})

describe('Dialog.Footer', () => {
  it('should match snapshot', () => {
    expect(render(<Dialog.Footer />).asFragment()).toMatchSnapshot()
  })
})
