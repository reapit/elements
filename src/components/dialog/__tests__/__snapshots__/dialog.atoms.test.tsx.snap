// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog.Body should match snapshot 1`] = `
<DocumentFragment>
  <mock-styled.div />
</DocumentFragment>
`;

exports[`Dialog.Container should match snapshot 1`] = `
<DocumentFragment>
  <mock-styled.dialog />
</DocumentFragment>
`;

exports[`Dialog.Footer should match snapshot 1`] = `
<DocumentFragment>
  <mock-styled.div />
</DocumentFragment>
`;

exports[`Dialog.Title should match snapshot 1`] = `
<DocumentFragment>
  <mock-styled.h2 />
</DocumentFragment>
`;
